qiskit/transpiler/passes/optimization/light_cone.py

```diff
@@ -0,0 +1,137 @@
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.transpiler.passes.utils.remove_final_measurements import calc_final_ops
+
+commutator = CommutationChecker(standard_gates_commutations)
+translation_table = str.maketrans({"+": "X", "-": "X", "l": "Y", "r": "Y", "0": "Z", "1": "Z"})
+
+
+class LightCone(TransformationPass):
+    """Remove the gates that do not affect the outcome of a measurement on a circuit.
+
+    Pass for computing the light-cone of an observable or measurement. The Pass can handle
+    either an observable one would like to measure or a measurement on a set of qubits.
+    """
+
+    def __init__(self, bit_terms: str | None = None, indices: list[int] | None = None) -> None:
+        """
+        Args:
+            bit_terms: If ``None`` the light-cone will be computed for the set of measurements
+                in the circuit. If a string is specified, the light-cone will correspond to the
+                reduced circuit with the same expectation value for the observable.
+            indices: list of non-trivial indices corresponding to the observable in ``bit_terms``.
+        """
+        super().__init__()
+        valid_characters = {"X", "Y", "Z", "+", "-", "l", "r", "0", "1"}
+        self.bit_terms = None
+        if bit_terms is not None:
+            if not indices:
+                raise ValueError("`indices` must be non-empty when providing `bit_terms`.")
+            if not set(bit_terms).issubset(valid_characters):
+                raise ValueError(
+                    f"`bit_terms` should contain only characters in {valid_characters}."
+                )
+            if len(bit_terms) != len(indices):
+                raise ValueError("`bit_terms` must be the same length as `indices`.")
+            self.bit_terms = bit_terms.translate(translation_table)
+        self.indices = indices
+
+    @staticmethod
+    def _find_measurement_qubits(dag: DAGCircuit) -> set[Qubit]:
+        final_nodes = calc_final_ops(dag, {"measure"})
+        qubits_measured = set()
+        for node in final_nodes:
+            qubits_measured |= set(node.qargs)
+        return qubits_measured
+
+    def _get_initial_lightcone(
+        self, dag: DAGCircuit
+    ) -> tuple[set[Qubit], list[tuple[Gate, list[Qubit]]]]:
+        """Returns the initial light-cone.
+        If obervable is `None`, the light-cone is the set of measured qubits.
+        If a `bit_terms` is provided, the qubits corresponding to the
+        non-trivial Paulis define the light-cone.
+        """
+        lightcone_qubits = self._find_measurement_qubits(dag)
+        if self.bit_terms is None:
+            lightcone_operations = [(ZGate(), [qubit_index]) for qubit_index in lightcone_qubits]
+        else:
+            # Having both measurements and an observable is not allowed
+            if len(dag.qubits) < max(self.indices) + 1:
+                raise ValueError("`indices` contains values outside the qubit rage.")
```

Comment:
Though I like raging qubits, this seems to miss an "n" 😉 
```suggestion
                raise ValueError("`indices` contains values outside the qubit range.")
```